trie/trieNodesHandler.go

```diff
@@ -0,0 +1,63 @@
+package trie
+
+type trieNodesHandler struct {
+	hashesOrder   []string
+	existingNodes map[string]node
+	missingHashes map[string]struct{}
+}
+
+func newTrieNodesHandler() *trieNodesHandler {
+	return &trieNodesHandler{
+		hashesOrder:   make([]string, 0),
+		existingNodes: make(map[string]node),
+		missingHashes: make(map[string]struct{}),
+	}
+}
+
+func (handler *trieNodesHandler) addInitialRootHash(rootHash string) {
+	handler.missingHashes[rootHash] = struct{}{}
+	handler.hashesOrder = append(handler.hashesOrder, rootHash)
+}
+
+func (handler *trieNodesHandler) processMissingHashWasFound(n node, hash string) {
+	handler.existingNodes[hash] = n
+	delete(handler.missingHashes, hash)
+}
+
+func (handler *trieNodesHandler) jobDone() bool {
+	return len(handler.hashesOrder) == 0
+}
+
+func (handler *trieNodesHandler) replaceParentWithChildren(index int, parentHash string, children []node, missingChildrenHashes [][]byte) {
+	delete(handler.existingNodes, parentHash)
+
+	allChildrenHashes := make([]string, 0, len(children)+len(missingChildrenHashes))
+	for _, c := range children {
+		hash := string(c.getHash())
+		allChildrenHashes = append(allChildrenHashes, hash)
+		handler.existingNodes[hash] = c
+	}
+	for _, m := range missingChildrenHashes {
+		hash := string(m)
+		allChildrenHashes = append(allChildrenHashes, string(m))
+		handler.missingHashes[hash] = struct{}{}
+	}
+
+	handler.hashesOrder = replaceHashesAtPosition(index, handler.hashesOrder, allChildrenHashes)
+}
+
+func replaceHashesAtPosition(index int, initial []string, newData []string) []string {
```

Comment:
this method can be optimized

check the below suggestion:
```go
func replaceHashesAtPosition(index int, initial []byte, newData []byte) []byte {
	if index >= len(initial) || index < 0 {
		return initial
	}

	before := initial[:index]
	after := initial[index+1:]

	result := make([]byte, 0, len(before)+len(after)+len(newData))
	result = append(result, before...)
	result = append(result, newData...)
	result = append(result, after...)

	return result
}

func replaceHashesAtPosition2(index int, initial []byte, newData []byte) []byte {
	if index > len(initial) || index < 0 {
		return initial
	}

	after := initial[index:]
	//add some space
	initial = append(initial, newData...)
	copy(initial[index+len(newData):], after)
	copy(initial[index:], newData)
	return initial
}

func Test_replace(t *testing.T) {
	initial := []byte("hello world")
	newData := []byte("dummy ")
	index := 6
	numIterations := 100000

	startTime := time.Now()
	for i := 0; i < numIterations; i++ {
		initial = replaceHashesAtPosition2(index, initial, newData)
	}
	elapsed := time.Since(startTime)
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)
	fmt.Println("memStats", statistics.GetRuntimeStatistics())
	fmt.Println("time", elapsed)
}
```

with `replaceHashesAtPosition` which should be similar in operation as your method (I used byte array instead of string array) I get following results:
`memStats [timestamp 1664537308 num go 4 alloc 137.81 MB heap alloc 137.81 MB heap idle 31.16 MB heap inuse 140.19 MB heap sys 171.34 MB heap num objs 470021 sys mem 184.65 MB num GC 330]` 
`time 5.882921909s`

but with second method  `replaceHashesAtPosition2` I get:
`memStats [timestamp 1664537383 num go 4 alloc 28.88 MB heap alloc 28.88 MB heap idle 6.43 MB heap inuse 32.98 MB heap sys 39.41 MB heap num objs 530312 sys mem 48.33 MB num GC 4]`
`time 690.044901ms`

Reply:
done